nautobot/users/tests/test_api.py

```diff
@@ -73,6 +73,123 @@ def setUpTestData(cls):
+        ]
+
+    def test_provision_token_valid(self):
+        """
+        Test the provisioning of a new REST API token given a valid username and password.
+        """
+        data = {
+            "username": "user1",
+            "password": "abc123",
+        }
+        user = User.objects.create_user(**data)
+        url = reverse("users-api:token_provision")
+
+        response = self.client.post(url, data=data)
+        self.assertEqual(response.status_code, 201)
+        self.assertIn("key", response.data)
+        self.assertEqual(len(response.data["key"]), 40)
+        token = Token.objects.get(user=user)
+        self.assertEqual(token.key, response.data["key"])
+
+    def test_provision_token_invalid_username(self):
+        """
+        Test the behavior of the token provisioning view when a invalid username is supplied.
+        """
+        data = {
+            "username": "nonexistentuser",
+            "password": "abc123",
+        }
+        url = reverse("users-api:token_provision")
+
+        response = self.client.post(url, data=data)
+        self.assertEqual(response.status_code, 403)
+
+    def test_provision_token_invalid_password(self):
+        """
+        Test the behavior of the token provisioning view when a invalid password is supplied.
+        """
+        data = {
+            "username": "user1",
+            "password": "password",
+        }
+        User.objects.create_user(**data)
+        data = {
+            "username": "user1",
+            "password": "hunter2",
+        }
+        url = reverse("users-api:token_provision")
+
+        response = self.client.post(url, data=data)
+        self.assertEqual(response.status_code, 403)
+
+    # def test_provision_token_diff_user(self):
+    #     """
+    #     Test the behavior of the token provisioning view when a different user is supplied.
+    #     """
+    #     # Create a user and token
+    #     data = {
+    #         "username": "user1",
+    #         "password": "abc123",
+    #     }
+    #     user1 = User.objects.create_user(**data)
+    #     Token(user=user1).save()
+    #     token = Token.objects.get(user=user1)
+    #     # create a 2nd user we'll try to make a token for
+    #     data = {
+    #         "username": "user2",
+    #         "password": "password",
+    #     }
+    #     user2 = User.objects.create_user(**data)
+
+    #     header = {
+    #         "Authorization:": "Token " + token.key,
+    #     }
+    #     data = {
+    #         "user": user2.id.__str__(),
+    #         "expires": "2100-01-01T00:00:00.000Z",
+    #         "key": "abcdefghijklmnopqrstubwxyzabcdefghijklmnopqrst",
+    #         "write_enabled": "true",
+    #         "description": "string"
+    #     }
+    #     print(type(user2.id))
+    #     print(data)
+    #     url = reverse("users-api:tokens")
+
+    #     response = self.client.post(url, header=header, data=data)
+    #     self.assertEqual(response.status_code, 403)
```

Comment:
```suggestion
    def test_tokens_are_restricted_by_user(self):
        """
        Test that the tokens API can only access tokens belonging to the authenticated user.
        """
        # Create a user and token
        data = {
            "username": "user1",
            "password": "abc123",
        }
        self.user = User.objects.create_user(**data)
        Token(user=self.user).save()
        token = Token.objects.get(user=self.user)
        self.header = {
            "HTTP_AUTHORIZATION:": "Token " + token.key,
        }
        
        # List all tokens available to user1
        self.add_permissions("users.view_token")
        response = self.client.get(self._get_list_url(), **self.header)
        # Assert that only the user1_token appears in the results
        self.assertEqual(len(response.data["results"]), 1)
        self.assertEqual(response.data["results"][0]["id"], token.id)
        
        # Try to retrieve a token belonging to another user by ID
        response = self.client.get(self._get_detail_url(Token.objects.filter(user__n=self.user).first()), **self.header)
        self.assertHttpStatus(response, status.HTTP_404_NOT_FOUND)
```

Reply:
@glennmatthews Might be missing something in your suggestion, but where do you create a token by another user? As I read the suggestion you might be getting the 404 because no other tokens exist.

Reply:
In the base `APITestCase.setUp()`, one token is created for the default testuser, and in this test-case's `setUp()` two more tokens are created for the default test user. In my suggested test case here we then switch to a different user account (overwriting `self.user` because `self.add_permissions()` is much easier to use that way).

Reply:
Added this, couldn't get the `user__n` thing working (`django.core.exceptions.FieldError: Related Field got invalid lookup: n`) so just created a 2nd user.